vivisect/analysis/elf/elfplt_late.py

```diff
@@ -0,0 +1,376 @@
+    # and between PLT functions
+    lastva = pltva
+    jmpheur = {}
+    distanceheur = {}
+    for fva in curplts:
+        fsz = vw.getFunctionMeta(fva, 'Size')
+        gotva, gotsz = elfplt.getGOT(vw, fname)
+
+        offset = 0
+        while offset < fsz:
+            locva, lsz, ltype, ltinfo = vw.getLocation(fva + offset)
+            xrefsfrom = vw.getXrefsFrom(locva)
+            for xrfr, xrto, xrtype, xrtinfo in xrefsfrom:
+                if gotva <= xrto < gotva+gotsz:
+                    offcnt = jmpheur.get(offset, 0)
+                    jmpheur[offset] = offcnt + 1
+
+            offset += lsz
+
+        # capture distance between functions
+        delta = fva - lastva
+        distanceheur[delta] = distanceheur.get(delta, 0) + 1
+        lastva = fva
+
+    if None not in (gotva, gotsz):
+        logger.debug("GOT/size: 0x%x/0x%x", gotva, gotsz)
+
+    # GOT-XREF-Offset method
+    if len(jmpheur):
+        algo_GOT_XREF_Offset(vw, jmpheur, pltva, pltsz)
+
+    else:
+        logging.info("analyzePLT(0x%x, 0x%x) skipping GOT-XREF-Offset method: no existing functions found", pltva, pltsz)
+
+    # PLT-Func-Distance method
+    if len(distanceheur):
+        algo_PLT_Func_Distance(vw, curplts, distanceheur, pltva, pltsz)
+
+    else:
+        logging.info("skipping analyzePLT(0x%x, 0x%x) (PLT-Func-Distance method): no existing functions found", pltva, pltsz)
+
+    logger.info("elfplt_late (done): pltva: 0x%x, %d", pltva, pltsz)
+
+
+def algo_GOT_XREF_Offset(vw, jmpheur, pltva, pltsz):
+    # this seems to work for many PLT's, but only if the xref is identifiable 
+    # without function analysis.  it fails on i386-pic code which uses ebx 
+    # (which is handed into the call as a base address)
+
+    # if we have what we need... scroll through all the non-functioned area
+    # looking for GOT-xrefs
+    offbycnt = [(cnt, off) for off, cnt in jmpheur.items()]
+    offbycnt.sort(reverse=True)
+    logger.debug("PLT Segment Data: %r", vw.getSegment(pltva))
+    cnt, realoff = offbycnt[0]
+
+    # now roll through the PLT space and look for GOT-references from 
+    # locations that aren't in a function
+    logger.info("... analysis mode 1: GOT-XREF Offset")
+    offset = 0
+    while offset < pltsz:
+        locva, lsz, ltype, ltinfo = vw.getLocation(pltva + offset)
+
+        xrefsfrom = vw.getXrefsFrom(locva)
+        toGOT = False
+        for xrfr, xrto, xrtype, xrtinfo in xrefsfrom:
+            if isGOT(vw, xrto):
+                # make sure we're pointing at a valid import
+                toloc = vw.getLocation(xrto)
+                if toloc is None:
+                    continue
+
+                tlva, tlsz, tltype, tlinfo = toloc
+                if tltype not in (vivisect.LOC_POINTER, vivisect.LOC_IMPORT):
+                    continue
+
+                toGOT = True
+
+        logger.debug("loc: 0x%x   xrefs: %r (toGOT: %r)", locva, xrefsfrom, toGOT)
```

Comment:
What is this supposed to tell me?

Reply:
based on what the algorithm does, it tells you what location, what it's xrefs are, and whether it thinks they're in the GOT... which is what the algorithm is all about.

Reply:
since it will always follow a message stating the PLT Placement algorithm name, this is high data content without filling up your screen.  it's debug, so it should be ok for the details of the algorithm to be understood by the reader.  sorry for the lack of docstrings to make it easier, but that's rectified.